Correct the source location of methods and not null annotation checks [TG-8270]

majakusber opened the pull request:
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

https://github.com/diffblue/cbmc/pull/4802 introduced not-null-annotation checks for input parameters. These properties are added at the beginning of the goto program of the method. The source location of the property was missing information - line  number and method name. The information is inherited from the method itself, so as a fix I set the line and source location of the method itself.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

Commit:
Update method source location info

The line number is not the line number of the first instruction because
that can be unassigned. Instead, set it to the line number of the first
instruction that does have a line number assigned.

Set location function to be the name of the method itself.

Commit:
Update regression tests

Not-null annotation check properties which are added to the beginning of
the method's goto inherit the source_locationt of the method. Because
previously the location had missing information (missing method name),
the property name looked like this:
[not-null-annotation-check.1]

Now the name looks like this:
[java::Main.bar:(Ljava/lang/Integer;ZLjava/lang/Object;[Ljava/lang/String;)Ljava/lang/Integer;.not-null-annotation-check.1]

Commit:
Add regression test for Kotlin

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 8df4a6b20f7a9c744eef657e5cea63c739af46ed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118811498

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1256,10 +1256,23 @@ void java_bytecode_parsert::rmethod_attribute(methodt &method)
+    }
+
+    // method name
+    method.source_location.set_function(
```
Might be worth factoring out into a separate function as this code is duplicated here: https://github.com/diffblue/cbmc/blob/develop/jbmc/src/java_bytecode/java_bytecode_parser.cpp#L1254

majakusber replied:
Done, I moved the assignment to he method location higher up, and then just re-use it's value when assigning the instruction location.

peterschrammel commented on jbmc/src/java_bytecode/java_bytecode_parser.cpp:
```diff
@@ -1256,10 +1256,23 @@ void java_bytecode_parsert::rmethod_attribute(methodt &method)
 
-    // line number of method
+    // line number of method (the first line number available)
     if(!method.instructions.empty())
```
Isn't that covered by the check in line 1268?

majakusber replied:
Indeed, fixed

Commit:
Remove unnecessary check

If the vector is empty, std:find_if returns end

Commit:
Move and reuse source_location.function

If the assignment to the method.source_location.function is done first,
we can reuse it when assigning instruction.source_location.function

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 652484972222c9f5a39b948e7ed80392969a38e2).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118920380
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 85a1e8b3bcea7e78fd5515a83a847049f4f9753a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/118920462
